Sources/Resource/NetworkResource.swift

```diff
@@ -8,12 +8,47 @@
 
 import Foundation
 
-public typealias ResourceParseClosure<T> = (Data) throws -> T
+public protocol NetworkResource: Resource {
```

Comment:
I would perhaps break this protocol into simpler, more specific ones:

```swift

public protocol NetworkResource: Resource {

    var method: HTTP.Method { get }
    var headers: HTTP.Headers? { get }
    var query: HTTP.Query? { get }
    var body: Data? { get }

    func toRequest() -> URLRequest
}

public protocol StaticNetworkResource: NetworkResource {

    var url: URL { get }
}

public extension StaticNetworkResource {

    func toRequest() -> URLRequest {
        var urlRequest = URLRequest(url: url)

        urlRequest.allHTTPHeaderFields = headers
        urlRequest.httpBody = body
        urlRequest.httpMethod = method.rawValue

        return urlRequest
    }
}

public protocol RelativeNetworkResource: NetworkResource {

    static var baseURL: URL { get }

    var path: String { get }
}

public extension RelativeNetworkResource {

    func toRequest() -> URLRequest {
        var url = Self.baseURL

        if var components = URLComponents(url: url, resolvingAgainstBaseURL: false) {
            components.queryItems = buildQueryItems()
            components.path = components.path
                .appending(path)
                .replacingOccurrences(of: "//", with: "/")

            components.url.then {
                url = $0
            }
        }

        var urlRequest = URLRequest(url: url)

        urlRequest.allHTTPHeaderFields = headers
        urlRequest.httpBody = body
        urlRequest.httpMethod = method.rawValue

        return urlRequest
    }

    private func buildQueryItems() -> [URLQueryItem]? {
        guard let query = query, query.isEmpty == false else {
            return nil
        }

        return query.map { URLQueryItem(name: $0, value: $1) }
    }
}

```

This avoids having both `url` and `path` (and simultaneously optional), and makes the `toRequest` API more transparent.

The downside is that we might have to create more types (e.g. `StaticImageNetworkResource` and `RelativeImageNetworkResource`). However I think it may be worth it.

What do you think? 😝